services/snapshotter/service_test.go

```diff
@@ -1 +1,449 @@
+		t.Errorf("expected no message, got: %s", got)
+	}
+}
+
+func TestSnapshotter_RequestRetentionPolicyInfo(t *testing.T) {
+	s, l, err := NewTestService()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer l.Close()
+
+	var tsdbStore TSDBStore
+	tsdbStore.ShardFn = func(id uint64) *tsdb.Shard {
+		if id != 2 {
+			t.Errorf("unexpected shard id: %d", id)
+			return nil
+		}
+		return &tsdb.Shard{}
+	}
+	tsdbStore.ShardRelativePathFn = func(id uint64) (string, error) {
+		if id == 2 {
+			return "db0/rp0", nil
+		}
+		return "", fmt.Errorf("no such shard id: %d", id)
+	}
+
+	s.MetaClient = &MetaClient{Data: data}
+	s.TSDBStore = &tsdbStore
+	if err := s.Open(); err != nil {
+		t.Fatalf("unexpected open error: %s", err)
+	}
+	defer s.Close()
+
+	conn, err := net.Dial("tcp", l.Addr().String())
+	if err != nil {
+		t.Errorf("unexpected error: %s", err)
+		return
+	}
+	defer conn.Close()
+
+	req := snapshotter.Request{
+		Type:            snapshotter.RequestRetentionPolicyInfo,
+		Database:        "db0",
+		RetentionPolicy: "rp0",
+	}
+	conn.Write([]byte{snapshotter.MuxHeader})
+	enc := json.NewEncoder(conn)
+	if err := enc.Encode(&req); err != nil {
+		t.Errorf("unable to encode request: %s", err)
+		return
+	}
+
+	// Read the result.
+	out, err := ioutil.ReadAll(conn)
+	if err != nil {
+		t.Errorf("unexpected error reading database info: %s", err)
+		return
+	}
+
+	// Unmarshal the response.
+	var resp snapshotter.Response
+	if err := json.Unmarshal(out, &resp); err != nil {
+		t.Errorf("error unmarshaling response: %s", err)
+		return
+	}
+
+	if got, want := resp.Paths, []string{"db0/rp0"}; !reflect.DeepEqual(got, want) {
+		t.Errorf("unexpected paths: got=%#v want=%#v", got, want)
+	}
+}
+
+func TestSnapshotter_InvalidRequest(t *testing.T) {
+	s, l, err := NewTestService()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer l.Close()
+
+	if err := s.Open(); err != nil {
+		t.Fatalf("unexpected open error: %s", err)
+	}
+	defer s.Close()
+
+	conn, err := net.Dial("tcp", l.Addr().String())
+	if err != nil {
+		t.Errorf("unexpected error: %s", err)
+		return
+	}
+	defer conn.Close()
+
+	conn.Write([]byte{snapshotter.MuxHeader})
+	conn.Write([]byte(`["invalid request"]`))
+
+	// Read the result.
+	out, err := ioutil.ReadAll(conn)
+	if err != nil {
+		t.Errorf("unexpected error reading database info: %s", err)
+		return
+	}
+
+	// There should be no response.
+	if got, want := string(out), ""; got != want {
+		t.Errorf("expected no message, got: %s", got)
+	}
+}
+
+func NewTestService() (*snapshotter.Service, net.Listener, error) {
+	logger := zap.New(zap.NewTextEncoder())
+	s := snapshotter.NewService()
+	s.WithLogger(logger)
+
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// The snapshotter needs to be used with a tcp.Mux listener.
+	mux := tcp.NewMux()
+	go mux.Serve(l)
+
+	s.Listener = mux.Listen(snapshotter.MuxHeader)
+	return s, l, nil
+}
+
+type TSDBStore struct {
```

Comment:
You can DRY this up by using `internal.TSDBStoreMock{}`.

Reply:
Yea. I'll do that. I wrote this before that existed.